Add new event to be able to notify group creation.

gbastien opened the pull request:
Hi,

I created a new event "GroupCreated" and notify it in ZODBGroupManager.addGroup and DynamicGroupsPlugin.addGroup.

Could you please review this?  I will also make a PR for 1.x branch as we need a new release for Zope2.

Thank you!

Gauthier

gbastien committed:
Add new event to be able to notify group creation.

tseaver commented on Products/PluggableAuthService/plugins/DynamicGroupsPlugin.py:
```diff
@@ -331,6 +333,8 @@ def addGroup(self, group_id, predicate, title='', description='',
         self.ZCacheable_invalidate(view_name=view_name)
 
+        notify(GroupCreated(group_id))
+
```
Needs new test assertion that the event is emitted.

gbastien replied:
Hi @tseaver 

how can I do this? I can either register a testing subscriber that "does something", but I suppose there is another way to check this, do you have sample test somewhere?

Thank you for feedback,

Gauthier

icemac replied:
This could be an idea to write such a test.

gbastien replied:
Hi @tseaver @icemac 

I added a test using similar way as for UserFolder creation.  I finally had to add the group management plugin instance to the called event.  First it was to do the test work...  But finally I think it is useful to have a persistent object passed to the event or we only receive a group_id and we do not know where it comes from...

Please review and tell me.

Gauthier

tseaver commented on Products/PluggableAuthService/plugins/ZODBGroupManager.py:
```diff
@@ -180,6 +183,7 @@ def addGroup(self, group_id, title=None, description=None):
 
         self._groups[group_id] = {'id': group_id, 'title': title,
                                   'description': description}
+        notify(GroupCreated(group_id))
```
Needs new test assertion that the event is emitted.

tseaver reviewed (commented):
In addition to the new tests, please hold of on the 1.x branch PR until we get PR #21 landed.

icemac reviewed (changes requested):
LGTM, but it should have a test.

gbastien committed:
Added test showing that GroupCreated event is correctly emitted.
For convenience, we pass the PAS object to the event so we have something persistent to use because we only receive a group_id by default

gbastien committed:
Actually passing the plugin, not the PAS instance

icemac commented on Products/PluggableAuthService/plugins/tests/test_ZODBGroupManager.py:
```diff
@@ -78,6 +91,20 @@ def test_addGroup(self):
         self.assertEqual(info['id'], 'group')
 
+    def test_addGroup_CreatedEvent(self):
+        provideHandler(groupCreatedHandler)
```
This should be done in `setUp` and undone in `tearDown`.

gbastien replied:
@icemac mimics test_UserFolder.UserEvents.testUserCreationEvent...

I adapted code, thank you for review.

Gauthier

icemac reviewed (changes requested):
LGTM besides one request to change.

gbastien committed:
Register groupCreatedHandler in setUp and unregister it in tearDown

gbastien commented:
Hi @icemac @tseaver 

when merged, I will propose the same PR for the 1.x branch.

Thank you for feedback,

Gauthier

icemac commented:
@gbastien As you are a Zope committer you should be able to push the big merge button yourself.
